### DISCUSSION TITLE
transformStateForAlliance Reflecting Over Wrong Axis

### DISCUSSION THREAD

**Discussion by rutmanz** [mjansen4857/pathplanner#261] [Q&A] [answered]
I'm using `PathPlannerTrajectory.transformTrajectoryForAlliance` for my paths generated with PathWeaver, but it seems like they're being flipped across the wrong axis. I'm using Shuffleboard's Field2d widget and sending the trajectory there to visualize it, and the blue alliance looks fine, as expected
![blue](https://user-images.githubusercontent.com/92497727/221077322-62dcae81-e06f-46c1-b0a8-7b9f287a6ec4.png)

However, when transforming it for the red alliance, it flips to the upper half of the field, not the right half. 
![red](https://user-images.githubusercontent.com/92497727/221077147-4c5a3465-c4f9-4f7f-a03e-17fc24033c7d.png)

Looking at the method for transforming, it looks like it's being caused by this line in `transformStateForAlliance`. 
```java
Translation2d transformedTranslation =
          new Translation2d(state.poseMeters.getX(), FIELD_WIDTH_METERS - state.poseMeters.getY());
```
As I understand it, this should be
```java
Translation2d transformedTranslation =
          new Translation2d(FIELD_WIDTH_METERS -state.poseMeters.getX(),  state.poseMeters.getY());
```

Is this a bug or am I misunderstanding how it's meant to be used/what's going on here?

**Comment by hadley31** [comment, marked as answer]
Hi @rutmanz, this was answered in #241. But to summarize, the origin is always assumed to be right corner of your current alliance station, thus the transform is correct.

**Reply by rutmanz** [reply]
So is this a rendering issue with Shuffleboard not correctly rotating the origin that it uses?

**Reply by mjansen4857** [reply]
Shuffleboard will not change the origin of a Field2d widget. Take a look at [advantage scope](https://github.com/Mechanical-Advantage/AdvantageScope) if you want an odometry widget that will change origin based on alliance color.
